lib/cn/node_split.c

```diff
@@ -562,23 +569,40 @@ cn_split(struct cn_compaction_work *w)
     while (inflight > 0)
         usleep(100 * 1000);
 
-    for (i = 0; !err && i < w->cw_kvset_cnt; i++) {
+    for (int64_t i = w->cw_kvset_cnt - 1; !err && i >= 0; i--) {
         if (wargs[i].err) {
             err = wargs[i].err;
             break;
         }
 
         for (int k = LEFT; k <= RIGHT; k++) {
+            struct kvset_split_res *result = &wargs[i].result;
+            struct kvset_mblocks *blks = result->ks[k].blks;
             uint idx = (k == LEFT ? i : i + w->cw_kvset_cnt);
 
-            if (wargs[i].result.ks[k].blks->hblk.bk_blkid != 0) {
-                w->cw_kvsetidv[idx] = cndb_kvsetid_mint(cndb);
-                w->cw_split.dgen[idx] = wargs[i].ks->ks_dgen;
-                w->cw_split.compc[idx] = wargs[i].ks->ks_compc;
+            if (blks->hblk.bk_blkid != 0) {
+                if (drop_ptomb_ks[k] && blks->kblks.n_blks > 0)
+                    drop_ptomb_ks[k] = false;
+
+                if (HSE_UNLIKELY(drop_ptomb_ks[k])) {
+                    /* Drop contiguous kvsets containing only ptombs, starting from the oldest.
+                     */
+                    assert(blks->vblks.n_blks == 0);
+                    assert(result->ks[k].blks_commit->n_blks == 1);
+
+                    delete_mblock(w->cw_mp, &blks->hblk);
+                    blk_list_free(result->ks[k].blks_commit);
+                } else {
+                    w->cw_kvsetidv[idx] = cndb_kvsetid_mint(cndb);
+                    w->cw_split.dgen[idx] = wargs[i].ks->ks_dgen;
+                    w->cw_split.compc[idx] = wargs[i].ks->ks_compc;
+                }
```

Comment:
Bind the "drop contiguous..." comment to the code it's referring to (the delete_mblock() line...)